Lib/unittest/mock.py

```diff
@@ -2200,7 +2204,9 @@ def __init__(self, /, *args, **kwargs):
         self.__dict__['_mock_await_count'] = 0
         self.__dict__['_mock_await_args'] = None
         self.__dict__['_mock_await_args_list'] = _CallList()
-        code_mock = NonCallableMock(spec_set=CodeType)
+        code_mock = NonCallableMock(spec_set=_CODE_ATTRS)
```

Comment:
Can this be just `spec` with a list of string. There is a subtle difference between `spec` and `spec_set` that attributes can be set while using `spec`. Though we don't really expose `__code__` as part of the API this could just have the old behavior since someone might be using it in their tests and still have the performance benefits. Since we set `_spec_class` and `_spec_signature` I guess the validation for class and instantiation are also taken care of. 

Before patch

```
(myenv) ➜  cpython git:(main) ✗ ./python
Python 3.12.0a3+ (heads/main:0fe61d0838, Dec 18 2022, 11:39:41) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from unittest.mock import AsyncMock
>>> a = AsyncMock()
>>> a.__code__.a = 1 
```

After patch

```
(myenv) ➜  cpython git:(main) ✗ ./python
Python 3.12.0a3+ (heads/main:0fe61d0838, Dec 18 2022, 11:39:41) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from unittest.mock import AsyncMock
>>> a = AsyncMock()
>>> a.__code__.a = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/karthikeyan/stuff/python/cpython/Lib/unittest/mock.py", line 763, in __setattr__
    raise AttributeError("Mock object has no attribute '%s'" % name)
AttributeError: Mock object has no attribute 'a'
```

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock

> spec: This can be either a list of strings or an existing object (a class or instance) that acts as the specification for the mock object. If you pass in an object then a list of strings is formed by calling dir on the object (excluding unsupported magic attributes and methods). Accessing any attribute not in this list will raise an [AttributeError](https://docs.python.org/3/library/exceptions.html#AttributeError).
> If spec is an object (rather than a list of strings) then [__class__](https://docs.python.org/3/library/stdtypes.html#instance.__class__) returns the class of the spec object. This allows mocks to pass [isinstance()](https://docs.python.org/3/library/functions.html#isinstance) tests.

> spec_set: A stricter variant of spec. If used, attempting to set or get an attribute on the mock that isn’t on the object passed as spec_set will raise an [AttributeError](https://docs.python.org/3/library/exceptions.html#AttributeError).

Reply:
@tirkarthi How did you generate your "before patch" results? I cannot reproduce them. On latest `main` branch (git commit `797edb28c3dd02a5727f0374e937e906a389ab77`), this is what I see:

```
➜ ./python
Python 3.12.0a3+ (heads/main:797edb28c3, Dec 19 2022, 12:06:46) [GCC 8.5.0 20210514 (Red Hat 8.5.0-17)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from unittest.mock import AsyncMock
>>> a = AsyncMock()
>>> a.__code__
<NonCallableMock spec_set='code' id='140680885810624'>
>>> a.__code__.a = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/users/carljm/cpython/Lib/unittest/mock.py", line 763, in __setattr__
    raise AttributeError("Mock object has no attribute '%s'" % name)
AttributeError: Mock object has no attribute 'a'
```

And I see exactly the same results with my patch.

This is what I would expect, given that both the previous code and my patch use `spec_set` here, the only difference is that now I pass a list of strings to it instead of an object.

Reply:
Thanks @carljm , I was applying the patch manually and testing out a few things by manually editing code. I used spec by mistake during those edits and also overlooked the GitHub diff too. I verified spec_set is used from the initial AsyncMock commit. Apologies for the confusion. Please ignore the review.